transport/server.cc

```diff
@@ -601,12 +601,10 @@ future<> cql_server::connection::process_request() {
         auto stream = f.stream;
         auto mem_estimate = f.length * 2 + 8000; // Allow for extra copies and bookkeeping
         if (mem_estimate > _server._max_request_size) {
-            return _read_buf.skip(f.length).then([length = f.length, stream = f.stream, mem_estimate, this] () {
-                write_response(make_error(stream, exceptions::exception_code::INVALID,
-                        format("request size too large (frame size {:d}; estimate {:d}; allowed {:d}", length, mem_estimate, _server._max_request_size),
-                        tracing::trace_state_ptr()));
-                return make_ready_future<>();
-            });
+            write_response(make_error(stream, exceptions::exception_code::INVALID,
+                    format("request size too large (frame size {:d}; estimate {:d}; allowed {:d}", f.length, mem_estimate, _server._max_request_size),
+                    tracing::trace_state_ptr()));
+            return _read_buf.close();
```

Comment:
We have a bug/feature in Seastar's SSL (https://github.com/scylladb/seastar/issues/906, CC @elcallio) where _read_buf.close() **also** closes its output stream. So I don't know what `write_response` does - but if it doesn't wait for the write (and just puts it the response in some buffer, for example), the response will be lost if using SSL.

To test this with cql-pytest, we would need to write a test with an oversized request to elicit this, and then add "--https" support to cql-pytest (like we have in Alternator - I just sent a patch to change test/alternator/run to Python, so it's easy to add the same ssl stuff also to cql-pytest).

Reply:
Another note on this Seastar issue: @elcallio says that the Seastar behavior is not a bug, it's a feature, and it should stay. Let's assume for now that it does, so we need to work around it.

I think the easiest thing you can do is to wait for `_ready_to_respond` (when write_response completes) before calling `_read_buf.close()`. It think this will completely solve this problem, even if Seastar doesn't change.

Reply:
Right, thanks. For starters I'll reproduce the problem manually and see if waiting for `_ready_to_respond` helps.

Reply:
> Right, thanks. For starters I'll reproduce the problem manually and see if waiting for `_ready_to_respond` helps.

Because this is a race, it might not always reproduce but if you add a sleep in the write_response() it should be easier to reproduce. Also remember the problem only exist for SSL, not in normal TCP (where closing the read side doesn't close the writing side).